File: code/FeatureSelection.py
from skrebate import ReliefF
from sklearn import preprocessing
import numpy as np
import matplotlib.pyplot as plt

# ======================================================================================================================
if __name__ == '__main__':
    features_filename = "./results/main/reviews_features.csv"
    labels_filename = "./results/main/reviews_labels.csv"
    nFeatures_factor = 0.3
    n_neighbors = 10
    nFeatures = [50, 70, 100, 30, 15, 7]  # number of features for each feature type
    feature_names = ['unigram hist', 'bigram hist', 'trigram hist', 'fourgram hist', 'fivegram hist', 'LM perplexity']

    X = np.loadtxt(features_filename, delimiter=",")
    X = preprocessing.StandardScaler().fit_transform(X)
    labels = np.loadtxt(labels_filename, delimiter=",", dtype='str')
    class_to_labels_dict = list(set(labels))
    labels_to_class_dict = {label: i for i, label in enumerate(class_to_labels_dict)}
    y = np.array([labels_to_class_dict[label] for label in labels])

    n_features = int(X.shape[1] * nFeatures_factor)
    r = ReliefF(n_features_to_select=n_features, n_neighbors=n_neighbors)
    r.fit(X, y)

    feature_ranks = np.zeros(len(r.top_features_))
    for rank, ind in enumerate(r.top_features_):
        feature_ranks[ind] = rank

    colors = ['r', 'y', 'g', 'c', 'b', 'm', 'k']

    plt.figure()

    plt.subplot(2, 1, 1)
    ind = 0
    for i in range(len(nFeatures)):
        plt.bar(range(ind, ind + nFeatures[i]), r.feature_importances_[ind:(ind + nFeatures[i])], label=feature_names[i], color=colors[i])
        ind = ind + nFeatures[i]
    plt.bar(range(ind, len(r.feature_importances_)), r.feature_importances_[ind:], label='other', color=colors[-1])
    plt.grid()
    plt.xlabel('Feature')
    plt.ylabel('Feature importance weight')
    plt.legend(loc='best', shadow=True, fancybox=True, framealpha=0.7)

    plt.subplot(2, 1, 2)
    ind = 0
    for i in range(len(nFeatures)):
        plt.bar(feature_ranks[ind:(ind + nFeatures[i])], r.feature_importances_[ind:(ind + nFeatures[i])],
                label=feature_names[i], color=colors[i])
        ind = ind + nFeatures[i]
    plt.bar(feature_ranks[ind:], r.feature_importances_[ind:], label='other', color=colors[-1])
    plt.axvline(n_features, color='r', ls='--')
    plt.grid()
    plt.xlabel('Feature')
    plt.ylabel('Feature importance weight')
    plt.legend(loc='best', shadow=True, fancybox=True, framealpha=0.7)

    plt.suptitle('Feature Importance')
    plt.tight_layout()
    plt.show()

File: code/FeatureSelection.m
features_filename = fullfile('results','reviewer_classification','toy_features.csv');
labels_filename = fullfile('results','reviewer_classification','toy_labels.csv');
k = 10; % number of neighbors for reliefF
nFeatures = [50, 70, 100, 30, 15, 7]; % number of features for each feature type
feature_names = {'unigram hist',...
                 'bigram hist',...
                 'trigram hist',...
                 'fourgram hist',...
                 'fivegram hist',...
                 'LM perplexity'};

X = csvread(features_filename);
y = csvread(labels_filename);

% X = [-1,2,3,4,5,3,6,2,7,3,6;...
%       -1,5,2,6,-1,3,5,2,5,-8,3;...
%       -1,2,-2,43,1,4,4,4,-3,-1,-1;...
%       -10,2,2,2,2,0,0,0,-3,2,2;...
%       1,2,0,-1,-1,-1,-1,-2,-3,-4,-5;...
%       1,2,1,2,3,4,5,6,-3,8,9;...
%       1,0,0,0,0,0,0,0,0,0,0];
% y = [1,1,1,1,0,0,0].';

[idx,weights] = relieff(X,y,k);

figure();
hold on;
ind = 1;
for i = 1:length(nFeatures)
    bar(ind:(ind+nFeatures(i)-1), weights(ind:(ind+nFeatures(i)-1)))
    ind = ind + nFeatures(i);
end
grid on;
title('Feature Importance');
xlabel('Feature');
ylabel('Feature importance weight');
legend(feature_names);
